### PR TITLE
Implement support to insert events

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/76729

This PR addds the support to add events to TEC REST API.
In so doing I've added the Organizer and Venue endpints as (look at the code) it's just convenient to do stuff just once.
This PR relies heavily on the work done in `common` here: https://github.com/moderntribe/tribe-common/pull/406
and refactors much of our code to leverage the WP REST API validation mechanism as much as we can.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | ! must be followed by a single space | [Archive_Event.php:162](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L162)
:x: | There must be a single space after the first parenthesis | [Archive_Event.php:211](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L211)
:x: | There must be a single space before the last parenthesis | [Archive_Event.php:211](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L211)
:x: | There must be a single space after the first parenthesis | [Archive_Event.php:514](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L514)
:x: | There must be a single space before the last parenthesis | [Archive_Event.php:514](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L514)
:x: | No space found after comma in function call | [Archive_Event.php:514](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L514)
:x: | Comma not allowed after last value in single-line array declaration | [Single_Event.php:237](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Event.php#L237)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:x: | Comma not allowed after last value in single-line array declaration | [Single_Organizer.php:75](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L75)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)
:x: | Comma not allowed after last value in single-line array declaration | [Single_Venue.php:45](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L45)
:x: | Logical operator &quot;and&quot; is prohibited; use &quot;&amp;&amp;&quot; instead | [Base.php:57](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/REST/V1/Validator/Base.php#L57)
:x: | ! must be followed by a single space | [Base.php:70](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/Validator/Base.php#L70)
:x: | There must be a single space after the first parenthesis | [Base.php:70](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/Validator/Base.php#L70)
:x: | There must be a single space before the last parenthesis | [Base.php:70](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/Validator/Base.php#L70)
:x: | There must be a single space after the first parenthesis | [Base.php:70](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/Validator/Base.php#L70)
:x: | There must be a single space before the last parenthesis | [Base.php:70](https://github.com/moderntribe/the-events-calendar/blob/e3f3af90b9129e1f125f2513f276eaa94397b7f6/src/Tribe/Validator/Base.php#L70)

> via jenkins/codesniffer-phpcs

**Comment by lucatume** [comment]
This is on hold as I've got some things left to do:
1. cover the endpoints with wpunit tests
2. add/update the swagger documentation
3. have a discussion about how we want to manage duplicates with others devs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/99217a8543cd6ea2b4e8f4fb0584d4b70398df67/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/99217a8543cd6ea2b4e8f4fb0584d4b70398df67/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/99217a8543cd6ea2b4e8f4fb0584d4b70398df67/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/99217a8543cd6ea2b4e8f4fb0584d4b70398df67/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Expected 1 space before &quot;=&gt;&quot;; 0 found | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Expected 1 space after &quot;=&gt;&quot;; 0 found | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Expected 1 space before &quot;=&gt;&quot;; 0 found | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Expected 1 space after &quot;=&gt;&quot;; 0 found | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:129](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Event.php#L129)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:132](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Event.php#L132)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:151](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Event.php#L151)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:181](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Event.php#L181)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/56c14c766cab7e791a017614d7dcba06915f18a3/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:x: | Expected 1 space before &quot;=&gt;&quot;; 0 found | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Expected 1 space after &quot;=&gt;&quot;; 0 found | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:479](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L479)
:x: | Expected 1 space before &quot;=&gt;&quot;; 0 found | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Expected 1 space after &quot;=&gt;&quot;; 0 found | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Each line in an array declaration must end in a comma | [Archive_Event.php:486](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Archive_Event.php#L486)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:129](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Event.php#L129)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:132](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Event.php#L132)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:151](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Event.php#L151)
:x: | Each line in an array declaration must end in a comma | [Single_Event.php:181](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Event.php#L181)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/b677a0d9fd31a99953132ed13558b87eea49d2a0/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)

> via jenkins/codesniffer-phpcs

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/e046ff4a523af8eeb9607d3a5e3d5fe73ae52618/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/e046ff4a523af8eeb9607d3a5e3d5fe73ae52618/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/e046ff4a523af8eeb9607d3a5e3d5fe73ae52618/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/e046ff4a523af8eeb9607d3a5e3d5fe73ae52618/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)

> via jenkins/codesniffer-phpcs

File: src/Tribe/REST/V1/Endpoints/Single_Organizer.php [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			'date'        => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_time' ) ),
+			'date_utc'    => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_time' ) ),
+			'organizer'   => array( 'required' => true, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'description' => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'status'      => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_post_status' ) ),
+			// Organizer meta fields
+			'phone'       => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'website'     => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_url' ) ),
+			'email'       => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'image'              => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_image' ) ),
```
Comment (bordoni) [inline src/Tribe/REST/V1/Endpoints/Single_Organizer.php]: Align

File: src/Tribe/REST/V1/Endpoints/Single_Venue.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	/**
+	 * Returns the content of the `args` array that should be used to register the endpoint
+	 * with the `register_rest_route` function.
+	 *
+	 * @return array
+	 */
+	public function GET_args() {
+		// TODO: Implement GET_args() method.
```
Comment (bordoni) [inline src/Tribe/REST/V1/Endpoints/Single_Venue.php]: Use `@todo`

File: src/Tribe/REST/V1/Endpoints/Single_Venue.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			'address'       => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'city'          => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'country'       => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'province'      => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'state'         => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'zip'           => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'phone'         => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'stateprovince' => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_string' ) ),
+			'website'       => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_url' ) ),
+			'image'              => array( 'required' => false, 'validate_callback' => array( $this->validator, 'is_image' ) ),
```
Comment (bordoni) [inline src/Tribe/REST/V1/Endpoints/Single_Venue.php]: align

File: src/Tribe/REST/V1/Validator/Base.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+		if ( ! empty( $venue['id'] ) ) {
+			return tribe_is_venue( $venue['id'] );
+		}
+
+		$request = new WP_REST_Request();
+		/** @var Tribe__Events__REST__V1__Endpoints__Linked_Post_Endpoint_Interface $venue_endpoint */
+		$venue_endpoint = tribe( 'tec.rest-v1.endpoints.single-venue' );
+
+		$request->set_attributes( [ 'args' => $venue_endpoint->POST_args() ] );
```
Comment (bordoni) [inline src/Tribe/REST/V1/Validator/Base.php]: PHP 5.2 compatibility.

File: src/Tribe/REST/V1/Validator/Base.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			$is_associative_array = is_array( $entry ) && ( array_values( $entry ) !== $entry );
+			if ( ! $is_associative_array ) {
+				return false;
+			}
+
+			$request = new WP_REST_Request();
+			/** @var Tribe__Events__REST__V1__Endpoints__Linked_Post_Endpoint_Interface $organizer_endpoint */
+			$organizer_endpoint = tribe( 'tec.rest-v1.endpoints.single-organizer' );
+
+			$request->set_attributes( [ 'args' => $organizer_endpoint->POST_args() ] );
```
Comment (bordoni) [inline src/Tribe/REST/V1/Validator/Base.php]: PHP 5.2 compatibility.

**Comment by lucatume** [comment]
Fixed issues, good for another look

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Organizer.php:16](https://github.com/moderntribe/the-events-calendar/blob/371de7dac2bce22f86f1998c6cb285f6e7cfdefa/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L16)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Organizer.php:26](https://github.com/moderntribe/the-events-calendar/blob/371de7dac2bce22f86f1998c6cb285f6e7cfdefa/src/Tribe/REST/V1/Endpoints/Single_Organizer.php#L26)
:warning: | Comment refers to a TODO task &quot;Implement get() method&quot; | [Single_Venue.php:20](https://github.com/moderntribe/the-events-calendar/blob/371de7dac2bce22f86f1998c6cb285f6e7cfdefa/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L20)
:warning: | Comment refers to a TODO task &quot;Implement GET_args() method&quot; | [Single_Venue.php:30](https://github.com/moderntribe/the-events-calendar/blob/371de7dac2bce22f86f1998c6cb285f6e7cfdefa/src/Tribe/REST/V1/Endpoints/Single_Venue.php#L30)

> via jenkins/codesniffer-phpcs

File: src/Tribe/REST/V1/Endpoints/Single_Venue.php (hunk 2 above)
Comment (borkweb) [inline src/Tribe/REST/V1/Endpoints/Single_Venue.php]: Same goes for the `TODOs` in the Single_Organizer.php file

File: src/Tribe/REST/V1/Validator/Interface.php [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+
+interface Tribe__Events__REST__V1__Validator__Interface  extends Tribe__Events__Validator__Interface {
```
Comment (borkweb) [inline src/Tribe/REST/V1/Validator/Interface.php]: Is this file needed at all?

**Comment by borkweb** [comment]
Wow. There's a lot here, but overall it looks pretty darn solid! I just have a question about the necessity of one file (`src/Tribe/REST/V1/Validator/Interface.php`). Otherwise, a very big 🐘 :dash: to this!

File: src/Tribe/REST/V1/Validator/Interface.php (hunk 6 above)
Comment (lucatume) [inline src/Tribe/REST/V1/Validator/Interface.php]: Not immediately, no.
My idea here is to make REST V1 classes depend on a REST V1 specific interface to grant flexibility in the future.
While it currently could add empty code it will leave us free to specialize the validator API and methods (in terms of signature) down the road keeping the context all in the same place.
